src/Godot/Nativescript.hs

```diff
@@ -448,3 +601,103 @@ registerSignal (RegSignal desc (signalName, signalArgs)) = do
       copyVariant (arrPtr `plusPtr` (n * opaqueSizeOf @GodotVariant)) vtPtr
       withVars xs (n + 1) arrPtr mtd
   withVars [] n arrPtr mtd = mtd (arrPtr, fromIntegral n)
+
+foreign import ccall "dynamic"
+  call_godot_class_constructor_ :: FunPtr (IO (Object)) -> IO (Object)
+
+new :: forall o. (Object :< o, Typeable o, AsVariant o) => IO (Maybe o)
+new = do
+  con <- Foreign.withCString (T.unpack $ nameOf @o) (\str -> godot_get_class_constructor str)
+  if con == nullFunPtr then
+    pure Nothing else
+    do
+      o <- call_godot_class_constructor_ con
+      tryCast o
+
+newNativeScript :: forall a. NativeScript a => IO (Maybe a)
+newNativeScript = do
+  let name = nameOf @a
+  Just nativescript <- new @GApi.NativeScript
+  Just gdnlib <- tryCast @GApi.GDNativeLibrary =<< readIORef gdNativeLibraryRef
+  GN.set_library nativescript gdnlib
+  GN.set_class_name nativescript =<< toLowLevel name
+  no <- GN.new nativescript []
+  asNativeScript no
+
+emptyObject = Object nullPtr
+
+getNode :: forall b cls. (Object :< cls, GApi.Node :< cls,
+                    GApi.Node :< b, Typeable b, AsVariant b)
+        => cls -> Text -> IO b
+getNode self name = do
+  n :: GApi.Node <- get_node_or_null self =<< toLowLevel name
+  x <- tryCast n
+  case x of
+    Just r -> pure r
+    _ -> error "Error, getNode failed"
+
+getNodeNativeScript :: forall b child. (NativeScript b, GApi.Node :< child, Object :< child)
+         => child -> Text -> IO b
+getNodeNativeScript self name = do
+  (GApi.Node n) <- get_node self =<< toLowLevel name
```

Comment:
This brute-forced type casting is the bad practice enabled by the exposed type constructors and should be disallowed.

Reply:
Fixed in 56f4e80272503822e72085ef1e9b154e7f62be9b

Reply:
The old code is not a brute-force type cast. It is not a type cast at all.
I removed a lot of polymorphism from the library which makes type checking far more accurate. Just try to replace GApi.Node with some other type. You will get a type error.

The old code is correct because it's type checked. The new code just obfuscates this. It's no more polymorphic than the old code, but what type is being unwrapped is hidden and it's also needlessly slower.

This also shows you clearly why having type constructors is so good. The old code was correct, could never fail at runtime, and was crystal clear.
The new code is far more complex and looks as if it can fail at runtime. And indeed, with the new code, you can change it so that you can fail at runtime.

I would suggest undoing this and in general using a style that is much more explicit and has far less polymorphism. It's far easier to write code that way.

Reply:
Sorry, here you are right in that you're not casting, only deconstructing. I don't know why I picked this example, I wasn't thinking properly, but my point is that you can unsafely cast types by simply doing eg `Node actuallyJustAnObject` and there you go. That is absolutely bad practice.